### PR TITLE
Generate dependencies of preprocessed modules to avoid importing #ifdef import

### PR DESCRIPTION
**Pull request by Arthapz**

### CODE REVIEW

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 1]
```diff
@@ -152,8 +152,12 @@ function generate_dependencies(target, sourcebatch, opt)
             local jsonfile = path.translate(path.join(outputdir, path.filename(sourcefile) .. ".json"))
             common.fallback_generate_dependencies(target, jsonfile, sourcefile, function(file)
                 local compinst = target:compiler("cxx")
+                local defines = {}
+                for _, define in ipairs(target:get("defines")) do
+                    table.insert(defines, "-D" .. define)
+                end
                 local ifile = path.translate(path.join(outputdir, path.filename(file) .. ".i"))
-                os.vrunv(compinst:program(), {"-E", "-x", "c++", file,  "-o", ifile})
+                os.vrunv(compinst:program(), table.join(defines, {"-E", "-x", "c++", file,  "-o", ifile}))
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:160]: we need also support old msvc.

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 2]
```diff
@@ -150,7 +150,16 @@ function generate_dependencies(target, sourcebatch, opt)
 
             -- no support of p1689 atm
             local jsonfile = path.translate(path.join(outputdir, path.filename(sourcefile) .. ".json"))
-            common.fallback_generate_dependencies(target, jsonfile, sourcefile)
+            common.fallback_generate_dependencies(target, jsonfile, sourcefile, function(file)
+                local compinst = target:compiler("cxx")
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:154]:
```lua
        local clang, toolname = target:tool("cxx")
        assert(toolname == "clang")
        -- ...
        os.vrunv(clang, table.join(defines, {"-E", "-x", "c++", file,  "-o", ifile}))
```

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 3]
```diff
@@ -150,7 +150,16 @@ function generate_dependencies(target, sourcebatch, opt)
...
+                local defines = {}
+                for _, define in ipairs(target:get("defines")) do
+                    table.insert(defines, "-D" .. define)
+                end
+                local ifile = path.translate(path.join(outputdir, path.filename(file) .. ".i"))
+                os.vrunv(compinst:program(), table.join(defines, {"-E", "-x", "c++", file,  "-o", ifile}))
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:160]: two spaces, `,  "``

File: xmake/rules/c++/modules/modules_support/clang.lua (hunk 2 above)
Comment (waruqi) [reply xmake/rules/c++/modules/modules_support/clang.lua:154]: or we can use 

```lua
    local compinst = compiler.load("cxx", {target = target})
    local compflags = table.join("-E", "-x", "c++", compinst:compflags({target = target}))
    os.vrunv(compinst:program(), table.join(compflags, "-o", ifile))
```

and remove

```lua
local defines = {}
                for _, define in ipairs(target:get("defines")) do
                    table.insert(defines, "-D" .. define)
                end
```

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/clang.lua:154]: yeah but when some flags that are in compflags prevent -E from working (the compiler try to search module imports, we don't want that)

File: xmake/rules/c++/modules/modules_support/gcc.lua [hunk 4]
```diff
@@ -501,3 +510,13 @@ function get_depoutputflag(target)
     end
     return depoutputflag or nil
 end
+
+function get_cppversionflag(target)
+    local cppversionflag = _g.cppversionflag
+    if cppversionflag == nil then
+        local compinst = target:compiler("cxx")
+        local flags = compinst:compflags({target = target})
+        cppversionflag = table.find_if(flags, function(v) string.startswith(v, "-std=c++") end) or "-std=c++20"
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/gcc.lua:519]: _g.cppversionflag = cppversionflag

File: xmake/rules/c++/modules/modules_support/gcc.lua [hunk 5]
```diff
@@ -179,7 +179,16 @@ function generate_dependencies(target, sourcebatch, opt)
...
-                common.fallback_generate_dependencies(target, jsonfile, sourcefile)
+                common.fallback_generate_dependencies(target, jsonfile, sourcefile, function(file)
+                    local compinst = target:compiler("cxx")
+                    local defines = {}
+                    for _, define in ipairs(target:get("defines")) do
+                        table.insert(defines, "-D" .. define)
+                    end
+                    local ifile = path.translate(path.join(outputdir, path.filename(file) .. ".i"))
+                    os.vrunv(compinst:program(), table.join(defines, {get_cppversionflag(target), "-E", "-x", "c++", file,  "-o", ifile}))
+                    return io.readfile(ifile)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/gcc.lua:190]: you need remove it after reading

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 6]
```diff
@@ -729,3 +741,13 @@ function get_requiresflags(target, requires, opt)
         return requireflags
     end
 end
+
+function get_cppversionflag(target)
+    local cppversionflag = _g.cppversionflag
+    if cppversionflag == nil then
+        local compinst = target:compiler("cxx")
+        local flags = compinst:compflags({target = target})
+        cppversionflag = table.find_if(flags, function(v) string.startswith(v, "/std:c++") end) or "/std:c++latest"
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:750]: and here

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 7]
```diff
@@ -171,6 +171,9 @@ end
 
 -- generate dependency files
 function generate_dependencies(target, sourcebatch, opt)
+    local compinst = target:compiler("cxx")
+    local toolchain = target:toolchain("msvc")
+    local vcvars = toolchain:config("vcvars")
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:176]: we should pass `toolchain:runenvs()`

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/msvc.lua:176]: pass to what ?

Comment (waruqi) [reply xmake/rules/c++/modules/modules_support/msvc.lua:175]: > os.vrunv(compinst:program(), ..., {envs = vcvars})

here, 

```lua
local msvc = target:toolchain("msvc")

{envs = msvc()}
```

**Comment by Arthapz** [comment]
Ah i need to take care of dependency include directories, or the preprocessing will fail

**Comment by Arthapz** [comment]
it's seem we can't do anything more for clang 
https://github.com/llvm/llvm-project/issues/59287#issuecomment-1334019565

**Comment by waruqi** [comment]
thanks!

**Comment by Arthapz** [comment]
Maybe soon i can add support of p1689 for clang https://github.com/llvm/llvm-project/issues/57293#issuecomment-1254417171

**Pull request merged**